openai/embeddings_utils.py

```diff
@@ -12,6 +10,8 @@
 from tenacity import retry, stop_after_attempt, wait_random_exponential
 
 import openai
+from openai.datalib import numpy as np
+from openai.datalib import pandas as pd
```

Comment:
I wonder if we should call `assert_has_numpy` and `assert_has_pandas` in each function these modules are used so that it's very clear to users what to do to fix the issue (rather than getting a generic `'NoneType' object has no attribute` Python exception).

Reply:
The `embeddings_utis.py` module is not imported from anywhere and imports `sklearn` and other libraries listed in the `openai[embeddings]` extra.  I couldn’t find any docs, but its usage implies `pip install openai[embeddings]`, (which now also ensures numpy/pandas/etc.), so the experience of using `embeddings_utils.py` should be unchanged.

https://github.com/jakubroztocil/openai-python/blob/jakub/data-libraries-optional/setup.py#L46-L53

It could be improved, though. I think each optional extra — `embeddings`, `wandb`, and the new `datalib` — would deserve mention in the README.  I’ll add a section on the new one, and if you can give me some context on the other two, I’ll be happy to mention them too. 

I wasn't sure whether you’d be interested in the PR, but it looks like you are, so I’ll polish it a bit: I’m thinking maybe throwing an `ImportError` instead of just `Exception` from the `assert_has_*` functions, ensuring the error messages are clear, etc. 

It’s to a degree a backward-incompatible change (for existing users who don’t install `openai[embeddings]` and hit [this line](https://github.com/jakubroztocil/openai-python/blob/jakub/data-libraries-optional/openai/api_resources/embedding.py#L42)), so it might also be worth bumping the major version.

Reply:
Oh you're right this is an embeddings file so it will have the right dependencies.

Regarding the backward-incompatibility, yes it's unfortunate but personally I think it's probably ok as long as the error is clear and explains how to resolve the problem. Also the line in `read_any_format` is specific to embeddings so it's fine to assume that the embedding deps were installed.

See https://github.com/openai/openai-python/issues/124 for some historical context too about how deps have been handled too.